src/apps/vpn/models/servercountrymodel.cpp

```diff
@@ -364,10 +302,54 @@ void ServerCountryModel::retranslate() {
   endResetModel();
 }
 
+unsigned int ServerCountryModel::avgLatency() const {
+  if (m_numLatencySamples == 0) {
+    return 0;
+  } else {
+    return (m_sumLatencyMsec + m_numLatencySamples - 1) / m_numLatencySamples;
+  }
+}
+
 void ServerCountryModel::setServerLatency(const QString& publicKey,
                                           unsigned int msec) {
-  if (m_servers.contains(publicKey)) {
-    m_servers[publicKey].setLatency(msec);
+  if (!m_servers.contains(publicKey)) {
+    return;
+  }
+
+  Server& server = m_servers[publicKey];
+  if (server.latency() != 0) {
+    m_sumLatencyMsec -= server.latency();
+  } else {
+    m_numLatencySamples++;
+  }
+  m_sumLatencyMsec += msec;
+  server.setLatency(msec);
+
+  // Ugly iteration. Find the city for this server.
+  // TODO: Maybe it would be cleaner to move the latency data into ServerCity.
```

Comment:
Or move this servers into the cities.

Reply:
The problem is that we also have a need to frequently look up servers by their public key, so moving the `Servers` back into the `ServerCity` model  would just cause some ugly iteration elsewhere. What we almost need is something like a database where both servers and cities can be searched via multiple criteria.